diff -ru avrdude-5.8/confwin.c avrdude-5.8-serjtag/confwin.c
--- avrdude-5.8/confwin.c	Sat Mar  7 05:09:11 2009
+++ avrdude-5.8-serjtag/confwin.c	Sat Nov 28 12:26:22 2009
@@ -22,9 +22,11 @@
 
 #if defined(WIN32NATIVE)
 
+#include <stdio.h>
 #include <limits.h>
 #include <windows.h>
 
+extern char *progname;
 
 static char *filename;
 
@@ -32,7 +34,29 @@
 void win_sys_config_set(char sys_config[PATH_MAX])
 {
     sys_config[0] = 0;
-    
+#if 1
+{
+    char *confname;
+    char namebuf[PATH_MAX];
+    FILE *f;
+    int base_len = strlen(progname) - 4;
+
+    strcpy(namebuf, progname);
+
+    if (!strcasecmp(progname + base_len, ".exe")) {
+	strcpy(namebuf + base_len, ".conf");
+    } else {
+	strcpy(namebuf, ".conf");
+    }
+    /* Use Windows API call to search for the Windows default system config file.*/
+    SearchPath(NULL, namebuf, NULL, PATH_MAX, sys_config, &filename);
+    f = fopen(sys_config, "r");
+    if (f) {
+	fclose(f);
+	return;
+    }
+}
+#endif
     /* Use Windows API call to search for the Windows default system config file.*/
     SearchPath(NULL, "avrdude.conf", NULL, PATH_MAX, sys_config, &filename);
     return;
