--- avrdude-5.11/lexer.l	2013-03-30 15:21:00.000000000 +0900
+++ avrdude-5.11-serjtag/lexer.l	2013-03-30 15:24:25.000000000 +0900
@@ -120,6 +120,7 @@
 arduino          { yylval=NULL; return K_ARDUINO; }
 avr910           { yylval=NULL; return K_AVR910; }
 avr910_devcode   { yylval=NULL; return K_AVR910_DEVCODE; }
+serjtag          { yylval=NULL; return K_SERJTAG; }
 avrftdi          { yylval=NULL; return K_AVRFTDI; }
 usbasp           { yylval=NULL; return K_USBASP; }
 usbtiny          { yylval=NULL; return K_USBTINY; }
