diff -ru avrdude-5.8-confwin/usbasp.c avrdude-5.8/usbasp.c
--- avrdude-5.8-confwin/usbasp.c	2009-03-07 05:09:11.000000000 +0900
+++ avrdude-5.8/usbasp.c	2009-11-30 18:59:57.000000000 +0900
@@ -183,10 +183,36 @@
 
 static int usbasp_open(PROGRAMMER * pgm, char * port)
 {
+  char *p;
+  char vend_name[128];
+  char prod_name[128];
+
+  vend_name[0] = 0;
+  vend_name[127] = 0;
+  prod_name[0] = 0;
+  prod_name[127] = 0;
+
+  p = strchr(port,':');
+  if (p == port) {
+     strncpy(prod_name, port+1, 127);
+  } else if (p && p[1] == 0) {
+     strncpy(vend_name, port, 127);
+     vend_name[strlen(vend_name)-1] = 0;
+  } else if (p) {
+     strncpy(vend_name, port, 127);
+     p = strchr(vend_name,':');
+     if (p) {
+        *p = 0;
+        strcpy(prod_name, p+1);
+     }
+  }
+  if (!vend_name[0]) strcpy(vend_name, "www.fischl.de");
+  if (!prod_name[0]) strcpy(prod_name, "USBasp");
+
   usb_init();
 
-  if (usbOpenDevice(&PDATA(pgm)->usbhandle, USBASP_SHARED_VID, "www.fischl.de",
-		    USBASP_SHARED_PID, "USBasp") != 0) {
+  if ((usbOpenDevice(&PDATA(pgm)->usbhandle, USBASP_SHARED_VID, vend_name ,
+		    USBASP_SHARED_PID, prod_name) != 0)) {
 
     /* check if device with old VID/PID is available */
     if (usbOpenDevice(&PDATA(pgm)->usbhandle, USBASP_OLD_VID, "www.fischl.de",
@@ -213,12 +239,37 @@
 }
 
 
+/*
+ * parse the -E string
+ */
+static int usbasp_parseexitspecs(PROGRAMMER * pgm, char *s)
+{
+  char *cp;
+
+  pgm->exit_reset = EXIT_RESET_DISABLED; // default
+  while ((cp = strtok(s, ","))) {
+    if (strcmp(cp, "reset") == 0) {
+      pgm->exit_reset = EXIT_RESET_ENABLED;
+    }
+    else if (strcmp(cp, "noreset") == 0) {
+      pgm->exit_reset = EXIT_RESET_DISABLED;
+    }
+    else {
+      return -1;
+    }
+    s = 0; /* strtok() should be called with the actual string only once */
+  }
+
+  return 0;
+}
+
 static void usbasp_close(PROGRAMMER * pgm)
 {
   unsigned char temp[4];
-  memset(temp, 0, sizeof(temp));
-  usbasp_transmit(pgm, 1, USBASP_FUNC_DISCONNECT, temp, temp, sizeof(temp));
-
+  if (pgm->exit_reset == EXIT_RESET_ENABLED) {
+    memset(temp, 0, sizeof(temp));
+    usbasp_transmit(pgm, 1, USBASP_FUNC_DISCONNECT, temp, temp, sizeof(temp));
+  }
   usb_close(PDATA(pgm)->usbhandle);
 }
 
@@ -549,6 +600,7 @@
   pgm->close          = usbasp_close;
   pgm->read_byte      = avr_read_byte_default;
   pgm->write_byte     = avr_write_byte_default;
+  pgm->parseexitspecs = usbasp_parseexitspecs;
 
   /*
    * optional functions
